server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java

```diff
@@ -102,7 +102,8 @@ private void innerRun() throws Exception {
         final boolean isScrollSearch = context.getRequest().scroll() != null;
         final List<SearchPhaseResult> phaseResults = queryResults.asList();
         final SearchPhaseController.ReducedQueryPhase reducedQueryPhase = resultConsumer.reduce();
-        final boolean queryAndFetchOptimization = queryResults.length() == 1;
+        final boolean queryAndFetchOptimization = queryResults.length() == 1
+            && context.getRequest().searchType() == SearchType.QUERY_THEN_FETCH;
```

Comment:
Why would we not run the fetch phase and the query phase in the same roundtrip when the search type has a DFS phase?

Reply:
Previously, when there was a single shard we would always set the search type to `QUERY_THEN_FETCH`: https://github.com/elastic/elasticsearch/blob/8fb440dbf5bceb7df07bf80203834cc3eed82862/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L988-L992

This makes sense, since you don't need DFS with one shard. Now, we might still execute `DFS_QUERY_THEN_FETCH` with a single shard, when kNN is used. From failing tests, I saw that we don't apply the "query and fetch" optimization in the query phase when DFS is enabled. (Surprisingly, we use different codepath to execute the query phase when DFS is enabled vs. not. There are two different actions `QUERY_ID_ACTION_NAME` and `QUERY_ACTION_NAME`).

Here was my thinking for next steps:
* In a follow-up, I plan to use QUERY_THEN_FETCH for kNN with a single shard. This will remove the special handling, since we will always know QUERY_THEN_FETCH is used when there's one shard.
* For now I'll change this check to `queryResults.length() == 1 && context.getRequest().hasKnnSearch();` and add a comment.

Reply:
Thanks for explaining, the proposed follow-up sounds good to me.